Handle newed variables and add stack trace

jensli opened the pull request:
Makes the framework handle errors where `%ut` has been newed, and also print a stack trace or errors.

Output for an error when `%ut` has been newed before the change:

```
BADERROR -  throws an error on purpose
BADERROR^%utt5 -  throws an error on purpose - Error: 150373850,ERROR+4^%ut,%YDB-E-LVUNDEF, Undefined local variable: %ut("CHK")
-----------------------------------------------------------------------  [FAIL]
```

Note that because the error trap has an error the original error is not displayed.

Output for an error when `%ut` has been newed after the change:

```
BADERROR -  throws an error on purpose
BADERROR^%utt5 -  throws an error on purpose - Error: 150372778,BADERRORINNER+2^%utt5,%YDB-E-EXPR, Expression expected but not found
8: BADERRORINNER+2^%utt5  --   S X= ; syntax error on purpose
7: BADERROR+6^%utt5  --   D BADERRORINNER
6: EN1+80^%ut  --   . . D @%ut("ENT")
5: EN1+56^%ut  --   . F  S %utI=$O(%utETRY(%utI)) Q:%utI'>0  S %ut("ENUM")=%ut("ERRN")+%ut("FAIL") D
4: EN1+27^%ut  --   F  S %ut("CURR")=%ut("CURR")+1 Q:'$D(%utROU(%ut("CURR")))  S CURRROU=%utROU(%ut("CURR")) D  I $T(@("SHUTDOWN^"_CURRROU))'="" D @("SHUTDOWN^"_CURRROU)
3: EN+8^%ut  --   D EN1(.%utROU,%utLIST)
2: VERBOSE+3^%utt1  --   D EN^%ut($T(+0),VERBSITY) ; Run tests here, be verbose.
1: @  --  new $ETRAP,$zcmdline set $ETRAP=etrap DO VERBOSE^%utt1(1)
0: %XCMD+5^%XCMD  --   xecute "new $ETRAP,$zcmdline set $ETRAP=etrap "_%cli
-----------------------------------------------------------------------  [FAIL]
```

Note that the original error is displayed, and a helpful stack trace (only printed if run in verbose mode).

Commit:
Handle errors when %ut is newed

This makes the framework handle and display errors that happen in routines
that are called from a test but where the %ut variable is newed and thus
not available to the error trap code. In that case the error trap crashed
and did not display information about the original error.

This change updates the error trap to unwind the stack before handling the
error so that %ut is available and the original error can be displayed
properly to the user.

The BADERRORINNER is updated to test that this works as expected.

Commit:
Print stack trace on error

Updates the error trap to print a stack trace which shows where in the
code that the error occurred.

shabiel commented:
@jensli Thank you for the MR. I am pretty busy till the middle of next week. I will look at it then. A cursory look shows that things are mostly good.

Can you add a test for transactions for the error trap? You added a rollback.

jensli commented:
> @jensli Thank you for the MR. I am pretty busy till the middle of next week. I will look at it then.

No hurry.

> Can you add a test for transactions for the error trap? You added a rollback.

I made a force push with prettier solution which also removed the rollback. I thought I got in the force push before anyone had time to look at the code, but you were too fast.

I added the rollback command because I got transaction problems when I experimented earlier, "can't quit with ongoing transaction".

But in my latest tests I didn't find a way to reproduce those transaction error. I don't understand why I don't get them anymore. Error trap code it tricky.

So because of this I removed the rollback command. Let's leave it out for now.

shabiel commented on Routines/_ut.m:
```diff
@@ -494,3 +499,11 @@ D GETLIST(.%utROU,1,%ut("RSLT"))
  QUIT %ZH2
+ ;
+WTRACE ; Write a stack trace.
+ Q:$$GETSYS()'=47 ; Code only tested on GTM
```
This code should work fine on Cache. So you can remove this line.

jensli replied:
Done.

shabiel commented on Routines/_ut.m:
```diff
@@ -494,3 +499,11 @@ D GETLIST(.%utROU,1,%ut("RSLT"))
+ ;
+WTRACE ; Write a stack trace.
+ Q:$$GETSYS()'=47 ; Code only tested on GTM
+ Q:'%utVERB ; Disable if not verbose
```
Can you increment the version at the top of the file to 1.63?

jensli replied:
Done.

shabiel reviewed (commented):
Done with review. Just asking for a couple of minor changes, then okay to merge.

Commit:
Increment file version

jensli commented:
> Done with review.

What are you doing working on a Saturday Sam don't you people ever take time off!

> Just asking for a couple of minor changes, then okay to merge.

Done.
